File: apps/web/lib/auth/schemas.ts
import { z } from "zod";

export const roleSchema = z.object({
  id: z.string().uuid(),
  name: z.enum([
    "SUPER_ADMIN",
    "PLACEMENT_CELL",
    "PLACEMENT_CONVENER",
    "STUDENT",
  ]),
  description: z.string(),
});

export const userSchema = z.object({
  id: z.string().uuid(),
  college_email: z.string().email(),
  personal_email: z.string().email().nullable(),
  first_name: z.string(),
  last_name: z.string(),
  display_name: z.string(),
  profile_picture: z.string().nullable(),
  is_active: z.boolean(),
  roles: z.array(roleSchema),
  primary_role: z
    .enum(["SUPER_ADMIN", "PLACEMENT_CELL", "PLACEMENT_CONVENER", "STUDENT"])
    .nullable(),
  primary_role_label: z.string(),
  workspace_path: z.string().nullable(),
  needs_welcome: z.boolean(),
  last_login: z.string().nullable(),
  created_at: z.string(),
});

export const authTokensSchema = z.object({
  access_token: z.string(),
  refresh_token: z.string(),
  token_type: z.string(),
  user: userSchema,
  is_new_user: z.boolean().default(false),
});

export const messageResponseSchema = z.object({
  message: z.string(),
});

export const apiErrorSchema = z.object({
  message: z.string(),
  error_code: z.string().optional(),
  request_id: z.string().nullable().optional(),
});

export type UserSchema = z.infer<typeof userSchema>;

File: apps/api/app/modules/users/schemas.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, EmailStr, Field

from app.modules.users.models import USER_STATUS_ACTIVE, User
from app.modules.users.workspace import (
    resolve_primary_role_name,
    resolve_workspace_path,
    role_display_label,
)


class RoleResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    description: str


class UserResponse(BaseModel):
    id: uuid.UUID
    college_email: EmailStr
    personal_email: EmailStr | None
    first_name: str
    last_name: str
    display_name: str
    profile_picture: str | None
    is_active: bool
    roles: list[RoleResponse]
    primary_role: str | None
    primary_role_label: str
    workspace_path: str | None
    needs_welcome: bool
    last_login: datetime | None
    created_at: datetime

    @classmethod
    def from_user(cls, user: User) -> "UserResponse":
        primary = resolve_primary_role_name(user)
        return cls(
            id=user.id,
            college_email=user.college_email,
            personal_email=user.personal_email,
            first_name=user.first_name,
            last_name=user.last_name,
            display_name=user.display_name,
            profile_picture=user.avatar_url,
            is_active=user.status == USER_STATUS_ACTIVE,
            roles=[RoleResponse.model_validate(role) for role in user.roles],
            primary_role=primary,
            primary_role_label=role_display_label(primary),
            workspace_path=resolve_workspace_path(user),
            needs_welcome=user.welcome_completed_at is None,
            last_login=user.last_login,
            created_at=user.created_at,
        )


class UserProfileUpdate(BaseModel):
    first_name: str | None = None
    last_name: str | None = None
    display_name: str | None = None
    avatar_url: str | None = None
    personal_email: EmailStr | None = None


class CreateUserData(BaseModel):
    college_email: EmailStr
    personal_email: EmailStr | None = None
    first_name: str = Field(min_length=1)
    last_name: str = Field(min_length=1)
    display_name: str = Field(min_length=1)
    avatar_url: str | None = None
    email_verified: bool = False
